### PR TITLE
Exception for multiple default languages #1849

### LINKED ISSUE #1849: Throw exception for multiple default languages
since i copy and pasted a language definition file and forgot to unset the default value i had two default languages in the language settings page. maybe did not cause any problems, but i was wondering what other parts of the kirby core or panel might do if there are two default languages. will it break? will only one be used? which one?

kirby 3.1.4

### PR DESCRIPTION
**Pull request by distantnative**
## Describe the PR
Throws a DuplicateException if multiple languages are set as default, so the user gets notified about faulty setup.

## Related issues
- Fixes #1849

## Todos
- [x] Add unit tests for fixed bug/feature
- [x] Pass all unit tests
- [x] Fix code style issues with CS fixer and `composer fix`

### CODE REVIEW

File: src/Cms/Languages.php [hunk 1]
```diff
@@ -45,11 +46,17 @@ public function create(array $props)
...
-        if ($language = $this->findBy('isDefault', true)) {
-            return $language;
-        } else {
-            return $this->first();
+        if ($defaults = $this->filterBy('isDefault', true)) {
+            if ($defaults->count() === 1) {
+                return $defaults->first();
+            }
+
+            if ($defaults->count() > 1) {
```
Comment (lukasbestle) [inline src/Cms/Languages.php]: That `if` clause is not necessary – the line will only be reached if the `if` clause above didn't match. So you can just throw the Exception here.

File: src/Cms/Languages.php [hunk 2]
```diff
@@ -45,11 +46,17 @@ public function create(array $props)
...
+                throw new DuplicateException('You cannot have two default languages. Please check your language config files.');
```
Comment (lukasbestle) [inline src/Cms/Languages.php]: Maybe "You cannot have multiple default languages" to cover all cases in the error message.

**Review by lukasbestle — CHANGES_REQUESTED** [review]
Besides these two smaller issues the changes LGTM. 👍

File: src/Cms/Languages.php (hunk 1 above)
Comment (distantnative) [reply src/Cms/Languages.php]: Are you sure? I thought `if ($defaults = $this->filterBy('isDefault', true))` might even be entered for empty collections. Or not?

Comment (lukasbestle) [reply src/Cms/Languages.php]: You are right, the `filterBy()` method actually always returns the collection object, which is always truthy (even if the collection is empty or the filter didn't find any values).

I think we should change it to something like this then:

```php
$defaults = $this->filterBy('isDefault', true);
if ($defaults && $defaults->isNotEmpty()) {
    if ($defaults->count() === 1) {
        return $defaults->first();
    }

    throw new DuplicateException('You cannot have multiple default languages. Please check your language config files.');
}

return $this->first();
```

**Comment by bastianallgeier** [comment]
I wonder if this is the right place to check for the duplicate. In my opinion we should do that when we load the languages.

**Comment by lukasbestle** [comment]
@bastianallgeier I think either solution is fine. The advantage of @distantnative's implementation is that we already have access to the languages collection and can filter it. If we'd check it when languages are loaded, we would have to use array functions.

**Comment by distantnative** [comment]
@bastianallgeier If we only do it when loading the languages, the exception won't be thrown if creating an `App` object by code with two default languages.

**Comment by lukasbestle** [comment]
@distantnative It would if we'd have the check in the `Languages` constructor, wouldn't it?

**Comment by distantnative** [comment]
I pushed a version that checks in the `Languages` constructor. Had to change how we initialize that class in two other parts to make it work.

**Review by lukasbestle — APPROVED** [review]
Actually looks a lot cleaner that way. 👍

**Review by bastianallgeier — APPROVED** [review]

**Comment by bastianallgeier** [comment]
Yep, that's a lot better.

**Pull request merged**